src/Rendering/applyPiecewiseFunctionOpacities.js

```diff
@@ -0,0 +1,43 @@
+function filterAndProcessOpacityNodes(numComps, opacityNodes) {
+  const filteredNodes = opacityNodes.map(node => node)
+  const baseLevel = 1.0 / numComps
+
+  for (let i = 0; i < filteredNodes.length; i++) {
+    if (filteredNodes[i].y < baseLevel) {
+      filteredNodes[i].y = baseLevel
+    }
+  }
+
+  const kernelSize = 9
+  const halfWindow = (kernelSize - 1) / 2
+
+  for (let i = 0; i < filteredNodes.length; i++) {
+    let sum = 0.0
+    let count = 0
+    for (let j = i - halfWindow; j <= i + halfWindow; j++) {
+      if (j > 0 && j < filteredNodes.length) {
+        sum += filteredNodes[j].y
+        count += 1
+      }
+    }
+    filteredNodes[i].y = sum / count
+  }
```

Comment:
For efficiency, can we do this loop for the starting and ending nodes (i.e. up to `halfWindow`, then something like:

```
const running = Array(kernelSize)
const sum = 0.0
for (let i = 0; i < kernelSize; i++) {
  running[i] = filteredNodes[i].y / kernelSize
  sum += running[i]
}

for (let i = halfWindowSize; i < filteredNodes.length -halfWindowSize; i++) {
  filteredNodes[i].y = sum
  sum -= running.shift()
  const next = filteredNodes[+1i].y / kernelSize
  running.push(next)
  sum += next
}
```

?